### DISCUSSION TITLE
Redundant code/confusion

### DISCUSSION THREAD

**Discussion by andyhasit** [parcel-bundler/parcel#4715] [General]
Hi, I read through the README and have found several issues on here which touch on the issue but I'm still not getting a clear explanation. I'm pretty new to this also.

I have written a library in ES6 and I have an app which uses that library, and am using parcel to build the app.

  - If the app imports directly from my local copy of the library (ES6>ES6) by relative path, parcel builds it fine.
  - But if I publish that library to npm, install it in my app, and import by package name (i.e. from node_modules) then Parcel complains.
  - If I transpile my library to ES5/Node modules using babel, it all works, except that the bundle produced by parcel has a lot of redundant code. Functions generated by babel such as `_classCallCheck` `_iterableToArray` `_arrayWithHoles` etc... all show up multiple times in the bundle.

I understand that parcel doesn't apply transformations to code imported from node_modules (unless you specify things somewhere?)

I understand that those functions show up because parcel essentially concatenates bundles from resources, and in this case those resources are files that were individually transpiled by babel, so each one of them has the polyfills it needs.

What I'm really not clear on is how to get around this:

1. Should I publish as ES6 and do something to ensure parcel treats it the same way as it would relative imports? (if so, how do I do this?)
2. Should I publish as ES5 but transpiling the library differently (e.g. to produce one file rather than one for each module, but in that case I will still get redundant code as `_classCallCheck` will be added once for the library and once for the app)

How do other projects deal with this? As you can see I'm a bit limited in my understanding of babel/node. 

Thanks :-)

**Comment by mischnic** [comment]
Ideally, you should publish your library as commonjs & esm.

```json
{
    "name": "my-library",
    "main": "dist-from-babel/index.js",
    "module": "src/index.js"
}
```
If your library doesn't use "real" ESM (so uses Flow/JSX/Typescript) you have to transpile it twice (to esm & cjs).

**Reply by andyhasit** [reply]
Thanks, but that didn't work... It causes the typical `TypeError: Class constructor xyz cannot be invoked without 'new'` which I presume is because parcel is still seeing this as a node_module and therefore not transpiling it.

**Reply by mischnic** [reply]
Adding a browserslist in the library should cause Parcel to transpile it anyway if the library's browserlist is older than the application's browserslist
